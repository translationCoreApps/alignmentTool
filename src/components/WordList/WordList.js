import React from 'react';
import PropTypes from 'prop-types';
import SecondaryToken from '../SecondaryToken';
import Token from 'word-map/structures/Token';
/**
 * Renders a list of words that need to be aligned.
 * Previously known as the "WordBank".
 * @param {Token[]} words,
 * @param {bool} isOver
 * @return {*}
 * @constructor
 */
const WordList = ({
  words,
  isOver
}) => {
  if (isOver) {
    return (
      <div
        style={{
          border: '3px dashed #44C6FF',
          height: '100%',
          width: '100%'
        }} />
    );
  } else {
    return (
      <React.Fragment>
        {words.map((token, index) => {
          return (
            <div
              key={index}
              style={{margin: '10px'}}>
              <SecondaryToken
                token={token}
                disabled={token.disabled === true}
              />
            </div>
          );
        })}
      </React.Fragment>
    );
  }
};

WordList.propTypes = {
  words: PropTypes.arrayOf(PropTypes.instanceOf(Token)).isRequired,
  isOver: PropTypes.bool.isRequired
};

export default WordList;
