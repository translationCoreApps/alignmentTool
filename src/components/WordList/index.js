import React from 'react';
import PropTypes from 'prop-types';
import {DropTarget} from 'react-dnd';
import * as types from '../WordCard/Types';
import WordList from './WordList';
import Token from 'word-map/structures/Token';

/**
 * Renders a word bank with drag-drop support
 */
class DroppableWordList extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      wordListScrollTop: null
    };
  }

  setScrollState(wordList, nextProps) {
    if (this.props.chapter !== nextProps.chapter || this.props.verse !== nextProps.verse) {
      wordList.scrollTop = 0;
      this.setState({wordListScrollTop: null});
    } else if (! this.props.isOver) {
      this.setState({wordListScrollTop: wordList.scrollTop});
    }
  }

  setWordListScroll(wordList) {
    if (! this.props.isOver && this.state.wordListScrollTop) {
      wordList.scrollTop = this.state.wordListScrollTop;
      this.setState({wordListScrollTop: null});
    }
  }

  componentWillReceiveProps(nextProps) {
    let wordList = document.getElementById('wordList');
    if(! wordList)
      wordList = this.props.wordList;
    this.setScrollState(wordList, nextProps);
  }

  componentDidUpdate() {
    let wordList = document.getElementById('wordList');
    if(! wordList)
      wordList = this.props.wordList;
    this.setWordListScroll(wordList);
  }

  render() {
    const {words, chapter, verse, connectDropTarget, isOver} = this.props;
    return connectDropTarget(
      <div
        id='wordList'
        style={{
          flex: 0.2,
          width: '100%',
          backgroundColor: '#DCDCDC',
          overflowY: 'auto',
          padding: '5px 8px 5px 5px'
        }}
      >
        <WordList
          chapter={chapter}
          verse={verse}
          words={words}
          isOver={isOver} />
      </div>
    );
  }
}

DroppableWordList.propTypes = {
  chapter: PropTypes.number,
  verse: PropTypes.number,
  words: PropTypes.arrayOf(PropTypes.instanceOf(Token)),
  connectDropTarget: PropTypes.func.isRequired,
  isOver: PropTypes.bool.isRequired,
  moveBackToWordBank: PropTypes.func.isRequired,
  wordList: PropTypes.object
};

/**
 * Handles drag events on the word bank
 */
const dragHandler = {
  drop(props, monitor) {
    props.moveBackToWordBank(monitor.getItem());
  }
};

const collect = (connect, monitor) => {
  return {
    connectDropTarget: connect.dropTarget(),
    isOver: monitor.isOver(),
    canDrop: monitor.canDrop()
  };
};

export default DropTarget(
  types.SECONDARY_WORD,
  dragHandler,
  collect
)(DroppableWordList);
