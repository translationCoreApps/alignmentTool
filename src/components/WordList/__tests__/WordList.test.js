/* eslint-env jest */

import {Token} from 'wordmap-lexer';
import React, {Component} from 'react';
import TestBackend from 'react-dnd-test-backend';
import {DragDropContext} from 'react-dnd';
import renderer from 'react-test-renderer';
import WordList from '../WordList';

describe('snapshot', () => {
  it('has no words', () => {
    const wrapper = renderer.create(
      <WordList
        isOver={false}
        modalOpen={false}
        words={[]}
        toolsSettings={{}}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('has words', () => {
    const WrappedWordList = wrapInTestContext(WordList);
    const w3 = new Token({text: 'w3'});
    w3.disabled = true;
    const words = [
      new Token({text: 'w1'}),
      new Token({text: 'w2'}),
      w3
    ];
    const wrapper = renderer.create(
      <WrappedWordList
        isOver={false}
        words={words}
        toolsSettings={{}}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
});


//
// Helpers
//

/**
 * Wraps a component into a DragDropContext that uses the TestBackend.
 */
function wrapInTestContext(DecoratedComponent) {
  return DragDropContext(TestBackend)(
    class TestContextContainer extends Component {
      render() {
        return <DecoratedComponent {...this.props} />;
      }
    }
  );
}
