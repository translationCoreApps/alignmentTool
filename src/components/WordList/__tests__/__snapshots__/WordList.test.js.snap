// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot has no words 1`] = `
<div
  style={
    Object {
      "height": "100%",
    }
  }
>
  <div
    style={
      Object {
        "display": "flex",
        "justifyContent": "flex-end",
        "padding": "0px 5px 5px",
      }
    }
  >
    <span
      className="glyphicon glyphicon-option-vertical"
      onClick={[Function]}
      style={
        Object {
          "cursor": "pointer",
          "fontSize": "18px",
          "margin": "0 0 0 10px",
          "padding": 0,
        }
      }
      title={undefined}
    />
  </div>
</div>
`;

exports[`snapshot has words 1`] = `
<div
  style={
    Object {
      "height": "100%",
    }
  }
>
  <div
    style={
      Object {
        "display": "flex",
        "justifyContent": "flex-end",
        "padding": "0px 5px 5px",
      }
    }
  >
    <span
      className="glyphicon glyphicon-option-vertical"
      onClick={[Function]}
      style={
        Object {
          "cursor": "pointer",
          "fontSize": "18px",
          "margin": "0 0 0 10px",
          "padding": 0,
        }
      }
      title={undefined}
    />
  </div>
  <div
    style={
      Object {
        "padding": "5px 10px",
      }
    }
  >
    <div
      onClick={[Function]}
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <div
        aria-describedby={null}
        className=""
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "flex": 1,
          }
        }
        title={null}
      >
        <div
          style={
            Object {
              "backgroundColor": "#FFFFFF",
              "borderLeft": "5px solid #44C6FF",
              "boxShadow": "0 1px 4px rgba(0, 0, 0, 0.3), 0 0 40px rgba(0, 0, 0, 0.1) inset",
              "cursor": "pointer",
              "display": "flex",
              "flexDirection": "row",
              "opacity": 1,
              "padding": "9px",
            }
          }
        >
          <span
            style={
              Object {
                "display": "flex",
                "flex": 1,
                "overflow": "hidden",
              }
            }
          >
            <span
              className=""
              onClick={[Function]}
              style={
                Object {
                  "flexGrow": 2,
                  "overflow": "hidden",
                  "textOverflow": "ellipsis",
                  "width": "max-content",
                }
              }
            >
              w1
            </span>
          </span>
        </div>
      </div>
    </div>
  </div>
  <div
    style={
      Object {
        "padding": "5px 10px",
      }
    }
  >
    <div
      onClick={[Function]}
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <div
        aria-describedby={null}
        className=""
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "flex": 1,
          }
        }
        title={null}
      >
        <div
          style={
            Object {
              "backgroundColor": "#FFFFFF",
              "borderLeft": "5px solid #44C6FF",
              "boxShadow": "0 1px 4px rgba(0, 0, 0, 0.3), 0 0 40px rgba(0, 0, 0, 0.1) inset",
              "cursor": "pointer",
              "display": "flex",
              "flexDirection": "row",
              "opacity": 1,
              "padding": "9px",
            }
          }
        >
          <span
            style={
              Object {
                "display": "flex",
                "flex": 1,
                "overflow": "hidden",
              }
            }
          >
            <span
              className=""
              onClick={[Function]}
              style={
                Object {
                  "flexGrow": 2,
                  "overflow": "hidden",
                  "textOverflow": "ellipsis",
                  "width": "max-content",
                }
              }
            >
              w2
            </span>
          </span>
        </div>
      </div>
    </div>
  </div>
  <div
    style={
      Object {
        "padding": "5px 10px",
      }
    }
  >
    <div
      onClick={[Function]}
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <div
        aria-describedby={null}
        className=""
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "flex": 1,
          }
        }
        title={null}
      >
        <div
          style={
            Object {
              "backgroundColor": "#FFFFFF",
              "borderLeft": "5px solid #868686",
              "boxShadow": "0 1px 4px rgba(0, 0, 0, 0.3), 0 0 40px rgba(0, 0, 0, 0.1) inset",
              "cursor": "not-allowed",
              "display": "flex",
              "flexDirection": "row",
              "opacity": 0.3,
              "padding": "9px",
              "userSelect": "none",
            }
          }
        >
          <span
            style={
              Object {
                "display": "flex",
                "flex": 1,
                "overflow": "hidden",
              }
            }
          >
            <span
              className=""
              onClick={[Function]}
              style={
                Object {
                  "flexGrow": 2,
                  "overflow": "hidden",
                  "textOverflow": "ellipsis",
                  "width": "max-content",
                }
              }
            >
              w3
            </span>
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
